### PR TITLE
🐛 Source Amazon Seller Partner: add `app_id` to  the OAuth Flow (source `spec` side)

### PR DESCRIPTION
**Pull request by bazarnov**
## What
Resolving: https://github.com/airbytehq/airbyte/issues/25773

## How
- edited `spec.py` by adding `app_id` to the `advanced_auth` specification to get this parameter from instance-wide params provided by Airbyte 
- removed `pydantic spec`  implementation and use the `spec.json` instead for easy modifications.
- corrected `test_source.py` unit tests
- updated `public docs` due to removed `app_id`

## 🚨 User Impact 🚨
This change is breaking since the `spec` has changed.

### CODE REVIEW

**Comment by bazarnov** [comment]
/publish connector=connectors/source-amazon-seller-partner pre-release=true run-tests=false

**Comment by bazarnov** [comment]
Pre-released version `1.2.1-dev.d38d6fba12` is published successfully.

**Comment by bazarnov** [comment]
The new `spec` without `app_id` has been tested here: https://github.com/airbytehq/airbyte-platform-internal/pull/7085

We can move on and merge this in, but I've faced other issues while `check_connection`. It turns out that `Role ARN` should also be supplied + `AWS access/secret keys`. Otherwise, we receive many meaningless config errors, even though we've passed the `OAuth Flow` correctly it's still required. @marcosmarxm why would it be these params are optional for the `source-amazon-seller-partner` considering this PR: https://github.com/airbytehq/airbyte/pull/23300

**Comment by bazarnov** [comment]
Changed GSM configs `start_date` to `2022-09-01` to guarantee `Orders` stream data presence.

**Comment by bazarnov** [comment]
edited by github-actions bot 
/test connector=connectors/source-amazon-seller-partner

File: airbyte-integrations/connectors/source-amazon-seller-partner/metadata.yaml
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-amazon-seller-partner/metadata.yaml:13]: Have we completed all the tests and can we leave the flag as true?

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:137]: Which argument do we set to `None`? And why should we consider changing the value of `sync_mode`?

**Review by lazebnyi — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:137]: The first arg is the `sync_mode`, it's mandatory, if `None` is passed to it the `full-refresh` is applied by default.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:136]: Not my change, what is the better naming you suggest?

**Comment by bazarnov** [comment]
@lazebnyi 
> Have we completed all the tests and can we leave the flag as true?

The OAuth is tested and working, I think we can leave this flag as `true` and get the connector back to Cloud if there are no more OAuth action items left.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:135]: Could you please be more specific with your question?

File: airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json:3]: I reuse it for the `incremental` test. Is there something wrong with updating this file?

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:137]: So why we need change this?

Comment (lazebnyi) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:137]: Using `sync_mode=SyncMode.full_refresh` is clearer and more explicit than setting it to silence None.

Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:137]: to simplify the unnecessary imports, no problem, I'll revert it back.

File: airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json:3]: I didn't find any references to this file in `acceptance-test-config`. Where you use this file?

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:137]: Anyway it is better to use a named argument when set to None or comment to be more clearer.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:136]: I'll leave it up to you to decide.

File: airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json:3]: Locally. Should I revert it back?)

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:135]: What is the `old config` ? 

if I understand you correctly, this is the input *.json config file? If so, the answer is yes, because the old config still requires the `app_id` that is no longer present, but was required. So this is the breaking change in either way.

File: airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json:3]: Yes

**Review by lazebnyi — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/sample_state.json]: Done

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py]: I've changed this to `stream_to_check`.

File: airbyte-integrations/connectors/source-amazon-seller-partner/metadata.yaml
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/metadata.yaml:13]: Yes, the results of the tests are available in secondary PR: https://github.com/airbytehq/airbyte-platform-internal/pull/7085

**Review by lazebnyi — APPROVED** [review]
LGTM!

**Review by pedroslopez — COMMENTED** [review]
Could you clarify why this is a breaking change? Is the change just removing a previously-required field? If so, it should be fine if this field is passed through an old config - am i missing something?

**Comment by bazarnov** [comment]
> Could you clarify why this is a breaking change? Is the change just removing a previously-required field? If so, it should be fine if this field is passed through an old config - am i missing something?

The field `app_id` is completely removed as the non-necessary field, it's just not used elsewhere but for OAuth purposes, which we do for the user with this change.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/Orders.json
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/Orders.json:1]: > how are you testing this on the dev envs? If doing so via connector version overrides in LaunchDarkly this shouldn't need to be set to true for this case.

I've deployed to `dev-2` and added the `pre-release` version mentioned above for the tests using `+New Connector` directly. 

I guess the `cloud.enabled: true` was not necessary, but I thought the process would be easier before the deployment took place, as missed the fact that the connector was hidden from the cloud. 

I didn't use `Launch Darkly`. I've made changes directly on the `dev-2` db with connectors' `source-definition-id` and was able to create and test the OAuth Flow for the `pre-release` custom connector.

> are these schema changes related to the spec change?

No. This eventually appeared after the `Orders` stream has finally enabled for the tests. It was not tested by CAT before.

Comment (pedroslopez) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/Orders.json:1]: > I didn't use Launch Darkly. I've made changes directly on the dev-2 db with connectors' source-definition-id and was able to create and test the OAuth Flow for the pre-release custom connector.

Ah got it- this worked because the connector is not on the cloud catalog. As a note, if something is in the catalog any manual changes made in the DB will get overwritten 30 seconds later by an automatic update that occurs.

Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/Orders.json:1]: Apparently, not for oauth_configs db  part)

**Comment by pedroslopez** [comment]
> The field `app_id` is completely removed as the non-necessary field, it's just not used elsewhere but for OAuth purposes, which we do for the user with this change.

Got it - this change in itself shouldn't be breaking. However, I'd like to understand what happens to users that have already authorized:
1. Do we have any users on cloud using this connector that have authorized with a manually provided app id?
2. If so, what happens to existing oauth authentications? would they stay connected given that we might be passing a different app_id?
3. Was the app_id only used for oauth? If not, does this have any effect on OSS users that would have passed app_id in their config?

**Comment by bazarnov** [comment]
> 1. Do we have any users on cloud using this connector that have authorized with a manually provided app id?

As far as this issue says - we do not: https://github.com/airbytehq/oncall/issues/2017#issuecomment-1563064033

> 2. If so, what happens to existing oauth authentications? would they stay connected given that we might be passing a different app_id?

Based on the answer in point 1. No. It was not working for 3rd-parties before, and half-working for those who tried using TCS's help. Since we provided our own `app_id` for this to happen, which will be resolved once we finally do this by default for OAuth Flow in Cloud.

> 3. It's required only for  OAuth, the OSS users are not affected. Tested.

File: airbyte-integrations/connectors/source-amazon-seller-partner/metadata.yaml
Comment (pedroslopez) [inline airbyte-integrations/connectors/source-amazon-seller-partner/metadata.yaml:5]: As far as I can tell this PR doesn't introduce any new features and just fixes an issue on our end. It also will not have any adverse effect on existing users. Following semver, this could be noted as a patch release rather than minor. Is there a special reason you went with a minor bump?

File: airbyte-integrations/connectors/source-amazon-seller-partner/acceptance-test-config.yml
Comment (pedroslopez) [inline airbyte-integrations/connectors/source-amazon-seller-partner/acceptance-test-config.yml:8]: curious, which backwards compatibility tests failed here? this shouldn't be noted as breaking

File: airbyte-integrations/connectors/source-amazon-seller-partner/acceptance-test-config.yml
Comment (pedroslopez) [inline airbyte-integrations/connectors/source-amazon-seller-partner/acceptance-test-config.yml:22]: This section is for breaking changes in the schema. This schema change just adds some missing fields to the orders - what was noted as breaking here?

**Review by pedroslopez — APPROVED** [review]
Did not functionally test this, but directionally looks good to me. Thanks for clarifying the breaking change questions - since this won't affect existing users I don't think we need any outreach here.

Left some questions on the acceptance test config to understand why our tests marked this as breaking. Non-blocking though.

We should make sure everything works with the changes outlined on the server-side PR (not hardcoding the app_id) before merging this in.

File: airbyte-integrations/connectors/source-amazon-seller-partner/metadata.yaml
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/metadata.yaml:5]: based on versioning rules, spec changes are considered "breaking", thus minor number was incremented.

File: airbyte-integrations/connectors/source-amazon-seller-partner/acceptance-test-config.yml
Comment (bazarnov) [reply airbyte-integrations/connectors/source-amazon-seller-partner/acceptance-test-config.yml:22]: the same comment as above about "app_id" filed.

**Comment by bazarnov** [comment]
Server-side PR: https://github.com/airbytehq/airbyte-platform-internal/pull/7085 has been updated and tested. Works as expected using `app_id` from `instance-wide` params.

**Comment by octavia-squidington-iii** [comment]
##  source-amazon-seller-partner test report (commit [`0d16ab009b`](https://github.com/airbytehq/airbyte/pull/27110/commits/0d16ab009b6ed402250d583f6ccf0fffc413c027)) - ✅

⏲️  Total pipeline duration: 2558 seconds

| Step                                                                                | Result   |
|:------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-amazon-seller-partner/metadata.yaml | ✅        |
| Connector version semver check.                                                     | ✅        |
| Connector version increment check.                                                  | ✅        |
| QA checks                                                                           | ✅        |
| Connector package install                                                           | ✅        |
| Build source-amazon-seller-partner docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                          | ✅        |
| Acceptance tests                                                                    | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5273453169)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-amazon-seller-partner test
```

**Comment by animer3009** [comment]
@bazarnov 
Where can I add app_id?
I have variable app_ids.
Early I was creating connectors using an API including app_id.
Now it pulls 0 rows after upgrading.
Urgent to solve that as I have that on production.
Please share some instructions...
I can't find where `advanced_auth` is...

**Comment by bazarnov** [comment]
@animer3009 Are you a Cloud user? why do you need the `app id`, it's not used elsewhere in the connector, therefore it was removed from the spec, as of it was needed just to obtain the access token. 

1) please share the screenshots from the ui you're using?
2) as for the emergency and you're using OSS, you're able to switch back to the previous version.
3) please share the logs of your latest sync with 0 records.

thank you

**Comment by animer3009** [comment]
@bazarnov
Nope, I am not the cloud user.

Now I custom builded connector. Returned back app_id.
Testing it right now....

I am using Airbyte 0.50.4
Screenshot at Jun 23 13-09-52
Screenshot at Jun 23 13-08-28
Screenshot at Jun 23 13-08-08
Screenshot at Jun 23 13-07-04


[b464d45a_4683_4c4c_8254_f9c72f5a24ed_logs_17482_txt.txt](https://github.com/airbytehq/airbyte/files/11850701/b464d45a_4683_4c4c_8254_f9c72f5a24ed_logs_17482_txt.txt)

I am not sure if downgrade only connector will help. First I will try custom builded connector.

**Comment by animer3009** [comment]
@bazarnov 
Same with custom build. 
Now I will try downgrade connector only. If it will not help, then will downgrade airbyte too..
I will keep you updated.

**Comment by bazarnov** [comment]
@animer3009 It's odd, because the last updates i've made to this connector didn't touch the fetch logic at all.

Could you verify whether or not any rate limits hit recently? as well as start date, please move it to a week ago, for instance.

I believe there is nothing to do with the `app_id` here.

**Comment by animer3009** [comment]
@bazarnov 
Possibly.
Yeah, downgrading Amazon Seller Partner to 1.2.0 does not help.
Then I downgraded Airbyte itself to the point where from I upgraded (1 release up) 0.44.12.
Before the upgrade it was 0.44.4...

Looks like it still started pulling... :) 
And yeah  in case of downgrading only the connector, it still finished in seconds. Filling was like it was not waiting for the ASP rapport generation.

My VM is production. I can't play with it, as it is still working, I will keep it as is. 
But I will try to make a test VM and play there.

**Comment by bazarnov** [comment]
@animer3009 I strongly recommend you to create an issue for this, and mention about the upgrading Airbyte, which caused problems, this should be investigated further. Thanks.

**Comment by animer3009** [comment]
@bazarnov 
UPDATE:
Unfortunately we were not able to repeat the issue on the test server.
We were not able to test all previous releases but reason possibly can be one of them.
One thing we did differently is that we keep our .ini and 
docker-compose.yaml file instead of pulling a fresh one.

P.s.
What I found during tests is that local build still include app_id optional input.
I am not sure if that is OK (just FYI).

Now we decided to still upgrade our production server.
We will observe stuff and keep you updated.

**Comment by animer3009** [comment]
@bazarnov 
same issue on production. looks like something wrong with API, as we create connector connections using Airbyte API on production.
Still troubleshooting.

**Comment by animer3009** [comment]
@bazarnov
I found the reason.
Looks like changed core logic and streams are not auto selected any more.
So I hardcoded selected = true on my end.
Now it is still working :)
Possibly guys who are using API calls will face the same issue. But don't think that there will be a lot of them lol :)

**Comment by bazarnov** [comment]
nicely done!

**Pull request merged**
